crates/red_knot_python_semantic/src/semantic_index/builder.rs

```diff
@@ -1018,6 +1136,44 @@ where
     }
 }
 
+#[derive(Debug, Default)]
+struct TryBlockContext {
+    snapshots: Vec<FlowSnapshot>,
+    visiting_nested_try_stmt: bool,
+}
+
+impl TryBlockContext {
+    fn push_snapshot(&mut self, snapshot: FlowSnapshot) {
+        self.snapshots.push(snapshot);
+    }
+}
+
+#[derive(Debug)]
+struct TryBlockContextsRefMut<'a>(RefMut<'a, Vec<TryBlockContext>>);
+
+impl<'a> TryBlockContextsRefMut<'a> {
+    fn current_try_block(&mut self) -> Option<&mut TryBlockContext> {
+        self.0.last_mut()
+    }
+}
+
+#[derive(Debug, Default)]
+struct TryBlockContexts(RefCell<Vec<TryBlockContext>>);
+
+impl TryBlockContexts {
+    fn push_try_block(&self) {
+        self.0.borrow_mut().push(TryBlockContext::default());
+    }
+
+    fn pop_try_block(&self) -> Option<TryBlockContext> {
+        self.0.borrow_mut().pop()
+    }
+
+    fn borrow_mut(&self) -> TryBlockContextsRefMut {
+        TryBlockContextsRefMut(self.0.borrow_mut())
+    }
+}
```

Comment:
So if I try applying this diff:

<details>

```diff
--- a/crates/red_knot_python_semantic/src/semantic_index/builder.rs
+++ b/crates/red_knot_python_semantic/src/semantic_index/builder.rs
@@ -96,7 +96,7 @@ impl<'db> SemanticIndexBuilder<'db> {
             .expect("Always to have a root scope")
     }
 
-    fn try_block_contexts(&self) -> &TryBlockContexts {
+    fn try_block_contexts(&mut self) -> &mut TryBlockContexts {
         self.try_block_contexts_stack.current_try_block_context()
     }
 
@@ -232,8 +232,7 @@ impl<'db> SemanticIndexBuilder<'db> {
             DefinitionCategory::Binding => use_def.record_binding(symbol, definition),
         }
 
-        if let Some(current_try_block) = self.try_block_contexts().borrow_mut().current_try_block()
-        {
+        if let Some(current_try_block) = self.try_block_contexts().current_try_block() {
             current_try_block.record_definition(self);
         }
 
@@ -780,9 +779,7 @@ where
                     .try_block_contexts()
                     .pop_try_block()
                     .expect("A `try` block should have been pushed to the stack prior to visiting the `body` field");
-                if let Some(parent_try_block) =
-                    self.try_block_contexts().borrow_mut().current_try_block()
-                {
+                if let Some(parent_try_block) = self.try_block_contexts().current_try_block() {
                     parent_try_block.record_visiting_nested_try_stmt();
                 }
                 self.flow_restore(pre_try_block_state);
@@ -877,9 +874,7 @@ where
                 // Account for the fact that we could be visiting a nested `try` block,
                 // in which case the outer `try` block must be kept informed about all the possible
                 // between-definition states we could have encountered in the inner `try` block(!)
-                if let Some(current_try_block) =
-                    self.try_block_contexts().borrow_mut().current_try_block()
-                {
+                if let Some(current_try_block) = self.try_block_contexts().current_try_block() {
                     current_try_block.record_exiting_nested_try_stmt();
                     current_try_block.record_definition(self);
                 }
diff --git a/crates/red_knot_python_semantic/src/semantic_index/except_handlers.rs b/crates/red_knot_python_semantic/src/semantic_index/except_handlers.rs
index de1918f71..b89bd4c52 100644
--- a/crates/red_knot_python_semantic/src/semantic_index/except_handlers.rs
+++ b/crates/red_knot_python_semantic/src/semantic_index/except_handlers.rs
@@ -1,5 +1,3 @@
-use std::cell::{RefCell, RefMut};
-
 use super::SemanticIndexBuilder;
 use super::use_def::FlowSnapshot;
 
@@ -11,9 +9,9 @@ impl TryBlockContextsStack {
         self.0.push(TryBlockContexts::default());
     }
 
-    pub(super) fn current_try_block_context(&self) -> &TryBlockContexts {
+    pub(super) fn current_try_block_context(&mut self) -> &mut TryBlockContexts {
         self.0
-            .last()
+            .last_mut()
             .expect("There should always be at least one `TryBlockContexts` on the stack")
     }
 
@@ -52,28 +50,19 @@ impl TryBlockContext {
     }
 }
 
-#[derive(Debug)]
-pub(super) struct TryBlockContextsRefMut<'a>(RefMut<'a, Vec<TryBlockContext>>);
-
-impl<'a> TryBlockContextsRefMut<'a> {
-    pub(super) fn current_try_block(&mut self) -> Option<&mut TryBlockContext> {
-        self.0.last_mut()
-    }
-}
-
 #[derive(Debug, Default)]
-pub(super) struct TryBlockContexts(RefCell<Vec<TryBlockContext>>);
+pub(super) struct TryBlockContexts(Vec<TryBlockContext>);
 
 impl TryBlockContexts {
-    pub(super) fn push_try_block(&self) {
-        self.0.borrow_mut().push(TryBlockContext::default());
+    pub(super) fn push_try_block(&mut self) {
+        self.0.push(TryBlockContext::default());
     }
 
-    pub(super) fn pop_try_block(&self) -> Option<TryBlockContext> {
-        self.0.borrow_mut().pop()
+    pub(super) fn pop_try_block(&mut self) -> Option<TryBlockContext> {
+        self.0.pop()
     }
 
-    pub(super) fn borrow_mut(&self) -> TryBlockContextsRefMut {
-        TryBlockContextsRefMut(self.0.borrow_mut())
+    pub(super) fn current_try_block(&mut self) -> Option<&mut TryBlockContext> {
+        self.0.last_mut()
     }
 }
```

</details>

Then the borrow check complains thusly:

```
error[E0502]: cannot borrow `*self` as immutable because it is also borrowed as mutable
   --> crates/red_knot_python_semantic/src/semantic_index/builder.rs:236:49
    |
235 |         if let Some(current_try_block) = self.try_block_contexts().current_try_block() {
    |                                          ---- mutable borrow occurs here
236 |             current_try_block.record_definition(self);
    |                               ----------------- ^^^^ immutable borrow occurs here
    |                               |
    |                               mutable borrow later used by call

error[E0502]: cannot borrow `*self` as immutable because it is also borrowed as mutable
   --> crates/red_knot_python_semantic/src/semantic_index/builder.rs:879:57
    |
877 |                 if let Some(current_try_block) = self.try_block_contexts().current_try_block() {
    |                                                  ---- mutable borrow occurs here
878 |                     current_try_block.record_exiting_nested_try_stmt();
879 |                     current_try_block.record_definition(self);
    |                                       ----------------- ^^^^ immutable borrow occurs here
    |                                       |
    |                                       mutable borrow later used by call
```

LMK if you can see a way of getting round this without the `RefCell`... I agree it's less than ideal, for sure!

Reply:
I'm fairly confident that there is a way to avoid the need for RefCell, and I also am interested in considering simpler representations than the current vec-of-vecs approach (e.g. can we use just a flat stack of try blocks, but via an enum that also has a no-try-block variant, so when we push scope we just push a no-try-block entry on the flat stack, and when we pop scope we pop it off?).

But I think this PR is complex enough that we should first make sure we have all the semantics and tests right, and _then_ worry about fine-tuning implementation strategy. So I suggest ignoring all implementation comments for now and focusing only on semantic comments.

(For example, the approach for removing the need for RefCell will look different depending on whether we do or don't need `visiting_nested_try_stmt`, which seems in question to me at the moment.)